src/lib/mongoBackend/dateExpiration.cpp

```diff
@@ -0,0 +1,198 @@
+* General Public License for more details.
+*
+* You should have received a copy of the GNU Affero General Public License
+* along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+*
+* For those usages not covered by this license please contact with
+* iot_support at tid dot es
+*
+* Author: Fermin Galan
+*
+*/
+#include <string>
+#include <vector>
+
+#include "common/string.h"
+#include "common/globals.h"
+#include "logMsg/logMsg.h"
+#include "ngsi/ContextAttribute.h"
+#include "parse/CompoundValueNode.h"
+#include "rest/OrionError.h"
+
+// FIXME P5: the following could be not necessary if we optimize the valueBson() thing. See
+// the next FIXME P5 comment in this file
+#include "mongoBackend/safeMongo.h"
+#include "mongoBackend/dbConstants.h"
+#include "mongoBackend/dateExpiration.h"
+
+
+
+/* ****************************************************************************
+*
+* USING
+*/
+using mongo::BSONObjBuilder;
+using mongo::BSONArrayBuilder;
+using orion::CompoundValueNode;
+
+
+
+/* ****************************************************************************
+*
+* getDateExpiration -
+*
+* Get the ISO8601 Expiration Date information for the given
+* ContextAttribute provided as parameter, in order to construct the corresponding Mongo Date_t object.
+*
+* It returns true, except in the case of error (in which in addition errDetail gets filled)
+*/
+static bool getDateExpiration
+(
+  const ContextAttribute*  caP,
+  mongo::Date_t*           dateExpiration,
+  std::string*             errDetail
+)
+{
+  if ((caP->type == DATE_TYPE) || (caP->type == DATE_TYPE_ALT))
+  {
+    *dateExpiration = mongo::Date_t(caP->numberValue*1000);
+
+    return true;
+  }
+
+  LM_E(("Runtime Error (attribute detected as date expiration but invalid type: %s)", caP->type.c_str()));
+  *errDetail = "error processing date expiration attribute, see log traces";
+
+  return false;
+}
+
+
+
+/* ****************************************************************************
+*
+* processDateExpirationAtEntityCreation -
+*
+* This function process the context attribute vector, searching for an attribute meaning
+* date expiration for the entity. In that case, it fills dateExpiration value.
+* If a date expiration attribute is not found dateExpiration keeps empty value set by the caller function.
+*
+* This function always return true (no matter if the attribute was found or not), except in an
+* error situation, in which case errorDetail is filled.
+*/
+bool processDateExpirationAtEntityCreation
+(
+  const ContextAttributeVector&  caV,
+  mongo::Date_t*                 dateExpiration,
+  std::string*                   errDetail,
+  OrionError*                    oe
+)
+{
+  for (unsigned ix = 0; ix < caV.size(); ++ix)
+  {
+    const ContextAttribute* caP = caV[ix];
+
+    if (caP->name != DATE_EXPIRES)
+    {
+      continue;
+    }
+
+    if (!getDateExpiration(caP, dateExpiration, errDetail))
```

Comment:
Just a little bit difficult to read ...
in one line it would be a lot easier (IMO):
```
if ((caP->name == DATE_EXPIRES) && !getDateExpiration(caP, dateExpiration, errDetail))
```
￼Also, once we have the attribute, and `getDateExpiration` worked, no real meaning to continue the loop, right?
=>
```
  if (caP->name == DATE_EXPIRES)
  ￼{
    if (!getDateExpiration(caP, dateExpiration, errDetail))
    {
      oe->fill(SccBadRequest, *errDetail, "BadRequest");
      return false;
￼    }

    return true;
  }

Reply:
Fixed in abdaa28